### ISSUE TITLE
Support rustup overrides properly

### ISSUE THREAD

**Issue by luser** [mozilla/sccache#87]
For the initial cut at Rust support we're not handling rustup overrides, since we're caching compiler info purely by compiler path. With rustup, the same compiler binary can invoke different Rust toolchains depending on the cwd, so we'll have to handle that somehow.

I was thinking of using a two-tier cache where we cache the "is this a Rust compiler?" result in the server code as we currently do, but then internally check for the presence of a `rustup` binary next to `rustc` and run `rustup which rustc` or something in the compile cwd to locate the actual compiler in use, and then have a cache of compilers using *that* path.

**Comment by alexcrichton** [comment]
cc @brson, you may have thoughts here as well

**Comment by drahnr** [comment]
Could it be possibly easier to resolve the compiler path, but instead hash the compiler binary itself and use that as key rather than the path? This would also remove all needs to duplicate the logic of `https://github.com/rust-lang/rustup/blob/master/src/config.rs`. How often would the hash need to be calculated?
The path is also not a good key, since it will differ from user to user.

**Comment by luser** [comment]
> The path is also not a good key, since it will differ from user to user.

The cache I'm discussing here is only in memory on the local machine where you're running sccache:
https://github.com/mozilla/sccache/blob/24f4306fad03ee77091d3cdd1520b78e9365076a/src/server.rs#L825-L865

For the hashes for compiled objects themselves we include hashes of the full contents of the Rust compiler's sysroot. Unfortunately we only do this the first time we locate a new compiler so this could result in invalid cache entries being stored when this situation is encountered.

**Comment by drahnr** [comment]
So the simplest version would be to track changes of the compiler sysroot (inotify, is there something similiar on windows?) and additionally check on launch of the server and invalidate the cache on each of those events, causing a full checksum of the sysroot to be done?
That should be enough?

**Comment by luser** [comment]
I think the best solution would still be something similar to what I proposed in my initial comment: make the Rust compiler cache two-tiered. The first tier would just be keyed by path+mtime, the second tier would check if this is a rustup rustc and if so use `rustup which rustc` in the cwd for the compile request to get the path to the actual compiler:
```
luser@eye7:/build/rust-minidump$ rustup which rustc
/home/luser/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/bin/rustc
```

Then the second tier could use *that* path+mtime to cache compiler info since that's the real compiler binary and will change if it gets updated. Additionally this would fix the original problem as stated where /src/a might compile with stable rust, /src/b might have an override for nightly rust, but they are both using the rustup shim `~/.cargo/bin/rustc` so sccache doesn't differentiate in the internal cache currently.

**Comment by drahnr** [comment]
@luser what's the preferred way of detecting `rustup`, currently only `which` is employed

**Comment by luser** [comment]
I would start with something simple: take the path to `rustc` and see if a `rustup` executable exists alongside it. Something like:
```rust
fn has_rustup(rustc: &Path) -> bool {
  rustc.with_file_name("rustup").with_extension(std::env::consts::EXE_EXTENSION).is_file()
}
```

(you might want something *slightly* fancier, the `which` crate sccache is already using does check that the file is executable etc.)

**Comment by drahnr** [comment]
I am currently using `which`, so that's all good then, thanks!

**Comment by drahnr** [comment]
It would be awesome if you could review #87 :)

**Comment by luser** [comment]
Fixed by #666.

**Comment by drahnr** [comment]
@luser there seems to be an uncovered edgecase which can still trigger this behaviour under certain unknown circumstances. I did not get around to investigate the issue yet, though it's been a few months (I can't seem to find the comments right now though).
Another point might be to create a release
